docs/proposals/20210126-nc-and-ncp.md

```diff
@@ -0,0 +1,304 @@
+- Define how each [component controller](/proposals/00_capn-glossary.md#component-controller) operates, this is defined in [Creating Control Plane Components](/proposals/20201026-creating-control-plane-components.md)
+
+
+## Proposal
+
+### User Stories
+
+1. As a control plane operator, I want to be able to set the namespace where my cluster is provisioned
+1. As a control plane operator, I want to be able to set the name of my control plane
+1. As a control plane operator, I want to be able to specify the [component controllers](/proposals/00_capn-glossary.md#component-controller) for my cluster.
+1. As a control plane operator, I want CAPN to manage my certificates.
+1. As a control plane operator, I want to supply my certificates and PKI.
+1. As a control plane operator, I want CAPN to manage my KUBECONFIGs.
+1. As a control plane operator, I want CAPN to setup object relationships.
+1. As a machine, I want to know how to address my control plane.
+
+### Components
+
+#### NestedCluster
+
+Kubernetes API Group: `infrastructure.cluster.x-k8s.io/v1alpha4`
+
+This resource is responsible for getting the Cluster API `Cluster` type and setting the `spec.controlPlaneRef` pointing back to to this `NestedCluster`, this allows the CAPI controllers to understand the state of the cluster. As well It listens to the `NestedControlPlane` resource which is associated and keeps it's status in sync as the cluster is provided.
+
+
+```go
+type NestedClusterSpec struct {
+	// ControlPlaneRef is an optional reference to a provider-specific resource that holds
+	// the details for provisioning the Control Plane for a Cluster.
+	// +optional
+	ControlPlaneRef *corev1.ObjectReference `json:"controlPlaneRef,omitempty"`
+	
+	// ControlPlaneEndpoint represents the endpoint used to communicate with the control plane.
+	// +optional
+	ControlPlaneEndpoint clusterv1.APIEndpoint `json:"controlPlaneEndpoint"`
+}
+
+type NestedClusterStatus struct {
+	// Ready is when the NestedControlPlane has a API server URL.
+	// +optional
+	Ready bool `json:"ready,omitempty"`
+}
+
+type NestedCluster struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   NestedClusterSpec   `json:"spec,omitempty"`
+	Status NestedClusterStatus `json:"status,omitempty"`
+}
+```
+
+### NestedCluster Lifecycle
+
+![Creating a NestedCluster](images/nestedcontrolplane/nc-activity.png)
+
+
+#### NestedControlPlane
+
+Kubernetes API Group: `controlplane.cluster.x-k8s.io/v1alpha4`
+
+The NestedControlPlane resource is responsible for orchestrating the overarching cluster, this controller doesn't create any of the downstream objects instead it creates a place where the downstream objects can look up shared values.
+
+```go
+type NestedControlPlaneSpec struct {
+	// ControlPlaneEndpoint represents the endpoint used to communicate with the control plane.
+	// +optional
+	ControlPlaneEndpoint clusterv1.APIEndpoint `json:"controlPlaneEndpoint"`
```

Comment:
This question may relate to the cluster-api. But why do we use APIEndpoint instead of `kubernetes service`? I noticed that the `APIEndpoint` contains only two fields, `host` and `port`. Does that mean if we wanna use service to expose the APIServer, we can simply set the `APIEndpoint.Host = <service-ip>`?

Reply:
Yea, that would be what we'd configure for the `ControlPlaneEndpoint`.